homeassistant/components/dhcp/__init__.py

```diff
@@ -55,13 +56,33 @@
 _LOGGER = logging.getLogger(__name__)
 
 
-class DhcpServiceInfo(TypedDict):
+@dataclass
+class DhcpServiceInfo:
     """Prepared info from dhcp entries."""
 
-    ip: str
+    ip: str  # pylint: disable=invalid-name
     hostname: str
     macaddress: str
 
+    # Used to prevent log flooding. To be removed in 2022.6
+    _warning_logged: bool = False
+
+    def __getitem__(self, name: str) -> Any:
+        """
+        Allow property access by name for compatibility reason.
+
+        Deprecated, and will be removed in version 2022.6.
+        """
+        if not self._warning_logged:
+            _LOGGER.debug(
+                "Accessing discovery_info['%s'] will fail in version 2022.6, "
+                "please use discovery_info.%s instead",
```

Comment:
Should we ask users to report these instances? And where?
As in, if we upgrade this to a warning later on, we should direct the user to the right place to get this resolved.

Reply:
Would that be just a standard link to https://github.com/home-assistant/core/issues ?
Or can/should we add HTTP-GET parameter helpers?
We'd also need the user to confirm the component.

Reply:
I think this is easiest:
```suggestion
                "Accessing discovery_info['%s'] will fail in version 2022.6, "
                "and should be replaced by discovery_info.%s. "
                "Please report this on https://github.com/home-assistant/core/issues",
```

Reply:
You could even do something like homeassistant/components/zeroconf/usage.py does to give the name of the integration

Reply:
I had a look at zeroconf usage but I couldn't work out how it worked of how I could apply it here 😞

Reply:
Something like this should do it.

```python3
from homeassistant.helpers.frame import report

...

        report(
            f"accessed discovery_info['{name}'] instead of discovery_info.{name}; This will fail in will fail in version 2022.6",
            error_if_core=False,
        )

````

Reply:
Aha! but that reports a warning and we don't want that right away.
Should I duplicate the report function? Or should I park it until 2021.12 is out (when we want to raise as a warning)?

Reply:
Or I can add an argument to `report` and `report_integration`:
 ```python
def report_integration(
    what: str,
    integration_frame: tuple[FrameSummary, str, str],
    level: int = logging.WARNING,
) -> None:
...
    _LOGGER.log(
        level,
        "Detected integration that %s. "
        ...
```

Reply:
My testing looks good :

```
DEBUG    homeassistant.helpers.frame:frame.py:94 Detected integration that accessed discovery_info['macaddress'] instead of discovery_info.macaddress; this will fail in version 2022.6. Please report issue for gogogate2 using this method at homeassistant/components/gogogate2/config_flow.py, line 51: await self.async_set_unique_id(discovery_info[dhcp.MAC_ADDRESS])
```